File: golang/graph_general/surrounded_regions.go
package main

func solve(board [][]byte) {
	if len(board) == 0 || len(board[0]) == 0 {
		return
	}

	rows := len(board)
	cols := len(board[0])

	var dfs func(r, c int)

	dfs = func(r, c int) {
		if r < 0 || r >= rows || c < 0 || c >= cols || board[r][c] != 'O' {
			return
		}

		board[r][c] = '#'

		dfs(r+1, c)
		dfs(r-1, c) // atas
		dfs(r, c+1) // kanan
		dfs(r, c-1) // kiri
	}

	for i := 0; i < rows; i++ {
		if board[i][0] == 'O' {
			dfs(i, 0)
		}
		if board[i][cols-1] == 'O' {
			dfs(i, cols-1)
		}
	}
	for j := 0; j < cols; j++ {
		if board[0][j] == 'O' {
			dfs(0, j)
		}
		if board[rows-1][j] == 'O' {
			dfs(rows-1, j)
		}
	}

	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if board[i][j] == 'O' {
				board[i][j] = 'X'
			} else if board[i][j] == '#' {
				board[i][j] = 'O'
			}
		}
	}
}

File: python/graph_general/surrounded_regions.py
from typing import List

class Solution:
    def solve(self, board: List[List[str]]) -> None:
        if not board or not board[0]:
            return


        rows, cols = len(board), len(board[0])

        def dfs(r, c):
            if r < 0 or r >= rows or c < 0 or c >= cols or board[r][c] != 'O':
                return

            board[r][c] = '#'

            dfs(r + 1, c)  # bawah
            dfs(r - 1, c)  # atas
            dfs(r, c + 1)  # kanan
            dfs(r, c - 1)  # kiri

        for i in range(rows):
            if board[i][0] == 'O':
                dfs(i, 0)

            if board[i][cols-1] == 'O':
                dfs(i ,cols - 1)

        for j in range(cols):
            if board[0][j] == 'O':
                dfs(0, j)
            if board[rows - 1][j] == 'O':
                dfs(rows - 1, j)


        for i in range(rows):
            for j in range(cols):
                if board[i][j] == 'O':
                    board[i][j] =  'X'
                elif board[i][j] == '#':
                    board[i][j] = 'O'
